### ISSUE TITLE
Determine locations of lighthouses.

### ISSUE THREAD

**Issue by cnlohr** [cnlohr/libsurvive#3]
Using data from the HMD, determine location of lighthouses relative to it.

**Comment by cnlohr** [comment]
This is some initial data.  Its columns are:
L, HED, [Actual Time (1/48 millionths of a second)], sensor ID, pulse code, delta from frame start.
[initial_hmd_data.txt.zip](https://github.com/cnlohr/libsurvive/files/644179/initial_hmd_data.txt.zip)

**Comment by Ardustorm** [comment]
Here is some code to visualize the points.
[viveStuff.zip](https://github.com/cnlohr/libsurvive/files/644193/viveStuff.zip)

**Comment by cnlohr** [comment]
Here's the data from my HMD data poins.

**Comment by cnlohr** [comment]
New tracking data - from one of the lighthouses, in mode 'A' pointed directly at the Vive.
[newtrackdata.csv.zip](https://github.com/cnlohr/libsurvive/files/644232/newtrackdata.csv.zip)

**Comment by cnlohr** [comment]
Fixed copy of initial hmd points...
[initial_hmd_data.txt.zip](https://github.com/cnlohr/libsurvive/files/644233/initial_hmd_data.txt.zip)

**Comment by cnlohr** [comment]
Here's the actual locations of all of the HMD points on the headset, in 3D space.

[hmd_points.csv.zip](https://github.com/cnlohr/libsurvive/files/644249/hmd_points.csv.zip)

**Comment by Ardustorm** [comment]
here is the data with some rudimentary animation so it is easier to see where everything is.
![animated](https://cloud.githubusercontent.com/assets/6853716/21078528/1c1aafea-bf28-11e6-8b8c-dd62d1f76406.gif)

**Comment by cnlohr** [comment]
Yeah, those normals don't look right :(

**Comment by ErrJordan** [comment]
Was the scad script edited and re-uploaded, because I just opened it to see if I could figure out why it looked strange and the swirling seems much less pronounced, and in the reverse direction.

This is a top view with orthogonal projection so it will look a little odd just due to the lack of perspective, but it gives a better sense of actual sensor orientation and they all look very nearly front-facing.  My intuition is that they shouldn't be and they should be more aligned with the curve of the surface, but it does make more sense than a swirl pattern.

![libsurvive](https://cloud.githubusercontent.com/assets/23353597/21082041/78d31432-bf98-11e6-9b8b-8231e24b8c40.png)

**Comment by Ardustorm** [comment]
Yeah, I've been trying to mess with it to show the normals. They don't look right because I just treated the vector as a rotation which is why the swirls appeared. I'm working on fixing that now.
ErrJordan, That might be because they are being rotated by less than one because of what I said above, thus it isn't noticeable.

**Comment by Ardustorm** [comment]
![vive](https://cloud.githubusercontent.com/assets/6853716/21083729/a4e868d4-bfab-11e6-9a22-e6a9f7b92362.gif)
Here is the corrected version, much better than last time.

And the corrected code here:
`   for(i =[0:31]) {
      color("red")translate(points[i]*100) {
	 x = normals[i][0];
	 y = normals[i][1];
	 z = normals[i][2];
	 rotate( [0,
	    1*atan2(sqrt(x*x + y*y),z),
	    1*atan2(y,x)]) {
	       cylinder(h= 1, r1=.25, r2=.1);
	    }
      }
   }`
[viveHeadsetVectors.zip](https://github.com/cnlohr/libsurvive/files/644836/viveHeadsetVectors.zip)

**Comment by Ardustorm** [comment]
What do the last two numbers on each line represent?
I made a spreadsheet of the data and found the average value for x and y for each sensor in the data-set. I find it interesting that sensor 10 does not have any Y data but does have X.
![screenshot from 2016-12-12 11-30-06](https://cloud.githubusercontent.com/assets/6853716/21113512/bccb7e88-c05e-11e6-9f53-1ed80810874c.png)
Here's the [Google Spreadsheet](https://docs.google.com/spreadsheets/d/1zDd8zhePRwxNPAZ0lxXIIqBeICiA6YbvLNc6svdTPcs/edit?usp=sharing) if anyone wants to look at all of the data

Here is another animation with each sensor labelled with it's number. The blue vectors represent the sensors with data from the newtrackdata.csv file. Point 10 is green because it had information only for X axis and wanted to see where it was located.
![vive_newtrackdata_labelled](https://cloud.githubusercontent.com/assets/6853716/21114813/44175d40-c063-11e6-9287-684803300cf7.gif)
[viveStuff.scad.zip](https://github.com/cnlohr/libsurvive/files/647064/viveStuff.zip)

**Comment by cnlohr** [comment]
The 0-3 represent the IR dat code, see https://github.com/nairol/LighthouseRedox/blob/master/docs/Light%20Emissions.md

The last column is the time, in 1/48 millionths of a second (aka ticks) of the time between the last sync pulse and the time that particular laser swept the point.

From what I can tell, 200,000 is dead center, it looks to be about 0.00009 degrees per tick, as I think one revolution is 400,000 ticks.  Someone should check my math on that, though.

**Comment by Ardustorm** [comment]
OK, thanks, so it looks like I don't need to worry about the IR dat code since you already calculated X and Y from it.
Here is the table I got from subtracting 200,000 from the number to center it and then scaling by **.000**9 to get degrees (400,000/360 = .0009 so it looks like you had an extra zero in there). 

X  (degrees from center) | Y (degrees from center)
------------ | -------------
-8.416500	|	-37.606250
-8.602045	|	-35.786689
-4.584530	|	-36.499129
-6.250571	|	-36.077187
-5.348712	|	-39.878528
-6.385500	|	-38.191597
-3.933958	|	
-7.979943	|	-39.510636
-0.641270	|	-34.350727
-1.718814	|	-35.612807
1.573604	|	-34.513354
2.023397	|	-36.402052
-0.798653	|	-39.376234
0.434340	|	-36.078383
0.306598	|	-38.124265

**Comment by cnlohr** [comment]
that totally looks believable, now comes figuring out the harder problem... Now, I'm curious if Alan Yates' twitter comment was in reference to us "Ugh this parametric surface interception is a multivariate transcendental equation, so much for a closed-form solution..."

I had given it some thought to try to linearize and turn into a matrix to solve, but, I don't know how well that would work.  I was planning on doing an iterative approach.  If anyone can find a better one, that would be super swanky.

I think I'm going to record more data, with IMUs, too.

**Comment by cnlohr** [comment]
Ben came through with the email to HTC!  They had our config!  Here is the config.
[correct_hmd_points.csv.zip](https://github.com/cnlohr/libsurvive/files/648096/correct_hmd_points.csv.zip)

**Comment by cnlohr** [comment]
Now, I have more info.  This is with two lighthouses pointed at the HMD on the floor.  I have also included IMU data for the HMD.
[two_lighthouses_test.csv.zip](https://github.com/cnlohr/libsurvive/files/648234/two_lighthouses_test.csv.zip)

**Comment by Ardustorm** [comment]
Cool, I'll have ti add the correct points in and compare, see how different they are to the one's we've been using.

Trying to solve the problem of finding the location of the lighthouses, I think I found something that might be useful. Since we know two points and the angle between them I was going to try and graph all possibilities (starting in 2d). I thought I got something wrong because I kept getting circles from the few points I put in. It turns out that [it is a property of circles](https://en.wikipedia.org/wiki/Circle#Inscribed_angles). My thoughts now are to try and come up with equations for 'spheres' for a few sets of points and see where they intersect. (I'll try to work on that tomorrow if I have time.)

**Comment by cnlohr** [comment]
vk2zay says we'll "have" to use iterative methods.  They're always my go-to and I was going to try to do one, myself... but now, I wanna try to find some other way just to be weird. :-D

**Comment by mwturvey** [comment]
Have you looked at the [Perspective-n-Point](https://en.wikipedia.org/wiki/Perspective-n-Point) problem?  If you consider the each lighthouse to be a camera, and the angular positions of each tracked point to be pixel locations of "points of interest" in an image that is generated by that "camera," you can start looking at the whole system as a more traditional computer vision problem.  At least that was a big "ah-ha" moment for me.
I'm probably repeating stuff you already know, but the problem you run into if you try and calculate the position information from a static set of data (i.e. a full x and y sweep) is that a moving tracked object will move to a different location between the time you get the X and Y sweeps.  So, unlike a traditional camera, you never get a simultaneous X and Y position. 
I *suspect* that the way it works is that an initial fix is acquired using one of the Perspective-n-Point problems.  Then, a Kalman filter is used to iteratively integrate sensor and IMU readings to maintain the position.  
I remember seeing a video at one point where Yates stated that they could maintain tracking for an already locked-on tracked object with a single sensor and IMU readings, as long as there was sufficient movement of the object.

**Comment by cnlohr** [comment]
The principle all along has been to use two different algorithms.  One to solve for camera position, then, during runtime, use camera positions to solve for positions and orientations of the objects.  These are different steps.  Right now, I'm only on the first step.  I strongly feel we should only focus on camera positions now.  I have my own ideas too about the runtime locations, but, that will have to wait until phase 3.

Re: PnP problem, that does look like the "right" way to solve it.  I have never heard of it, but I did once see that Hugin uses RANSAC to get camera positions and it is AMAZING at it.  Only thing is in Hugin's case it doesn't consider depth with its features so maybe this problem won't work out quite as AMAZINGly.

**Comment by Galastorm** [comment]
Following on from Mikes PnP comment, you may want to look at:
https://en.m.wikipedia.org/wiki/Epipolar_geometry
In theory, with enough samples, you should be able to reverse the base positions.

**Comment by cnlohr** [comment]
I don't know how you would invert that problem...  I would be interested to see if you have any specific ideas.  Re: the PnP route... someone has a C++ EPnP Solver in C++ -- but it is licensed under the FreeBSD license.  Since it's not part of the core algorithm, I might consider giving it a whirl: https://github.com/artivis/epnp

**Comment by Galastorm** [comment]
Sorry, the Vive also has a sensor in the base, which will give relative angles between the stations...

**Comment by mwturvey** [comment]
I believe that code is from the authors of the EPnP paper.  It requires OpenCV.  The problem I've run into, and haven't had time to get past yet is figuring out the intrinsic and extrinsic parameters of the Lighthouse "camera."  OpenCV has routines to calculate those for you, given a bunch of camera frames using one of two calibration images.  One calibration image is a circle, and the other is a checkerboard.  
The math got a little hairy for me, so I was hoping to use the OpenCV routines, at least to start with and get a proof-of-concept.  I think you could simulate a checkerboard pattern by constructing a custom sensor that is a grid of sensors on a flat board.  (Similar to [this](https://hackaday.io/project/15496/gallery#80e80eb78814674209c978f0bfa698b3), but in a planar grid.)  I think that calculating these parameters should be a one person, one time kind of thing.  Then, you should be able to use the EPnP algorithm to compute position.
Also, someone sufficiently knowledgeable in computer vision should be able calculate those parameters more or less "perfectly" without any custom hardware.

**Comment by Galastorm** [comment]
The Vive is a "perfect camera" system, without distortions or scaling issues.  You should be able to just implement it without the checkerboard calibration - I'm at work at the min I'll post a paper with a unity camera matrix later from home.
Using the Epipolar technique with the known angles between the stations, a single set of EPnP distances output should give you enough to "Fix" the base positions relative to each other.  You would then need to either use the IMU or "describe" the room like Vive in order to then fix the bases with respect to the ground plane.

**Comment by cnlohr** [comment]
@Galastorm: Any idea where that information could come from and/or how?  I didn't see any data coming back from the lighthouses.

@mwturvey: A few people have been worried about parameterizing the "cameras" in the vive system, but part of what makes it so magic is that they really are /perfect/.  It would take a herculean amount of effort to get regular cameras calibrated anywhere near as well as the lighthouses are naturally.  Also, you are correct, it does rely on OpenCV, which is not an acceptable dependency for libsurvive.  I wonder how hard it would be to write that out and keep the FreeBSD license vs. brew our own.

**Comment by Galastorm** [comment]
Sorry, no. I've just seen the sensor in a picture of a base teardown. I assumed (I know, ass-u and all that) that the information would be available somewhere in the system.
If you don't have direct access to the base angles, then they'd need to be worked out using simultaneous equations and sets of PnP angle readings.

**Comment by mwturvey** [comment]
I've not worked much with computer vision before, and couldn't find any info on what it would consider to be an ideal camera.  It sounds like it does mean that the angular distance between any two adjacent pixels is the same: that's awesome.  
I also would really like to have a small standalone library that can do this instead of a dependency on OpenCV.  My goal is to get it to run on a microcontroller-- basically to get high quality tracking in a small package.  There seemed to be a lot of tentacles once you get into the OpenCV code.  But, I think most of those tentacles com more from generic abstractions that OpenCV reasonably makes, but aren't needed in this algorithm. 
The approach I've been considering is starting with a small test app that could repeatedly run with each small change to see if anything broke, and surgically removing pieces out of OpenCV to get to a minimal implementation.

**Comment by cnlohr** [comment]
@mwturvey  Conveniently, the "where are my lighthouses" problem is something that would probably not be needed on the microcontroller.  Only the localization code.  That said.  Even for the PC.  OpenCV is unacceptable for this project.

I was just gonna take a whack at implementing the OpenCV functions they use, myself tonight.

**Comment by revirescam** [comment]
Here are some amateur's thoughs.

Background understanding (and the TSSP)
---------------------------------------
Timings are measured in "ticks" and are "since the last sync pulse", so I'll call them Ticks Since Sync Pulse" TSSPs.  I understand "sync pulses" to be wide angle flashes that don't scan, but flood the
room.  Subsequently horiz. and vert. scans are done with rotating lasers (or mirrors) that shine a sheet of light across the room at a known angular speed.  The TSSPs are generated by each photoreceptor on the headset detecting the the H/V scans.  This is the data the headset collects (part of it, at least, the IMU does accelerometer stuff too).

It's NOT a speed-of-light thing
-------------------------------

First - this is absolutely NOT a speed of light problem.  In acoustics you can tell with your ears where a source of sound is from the Time Difference Of Arrival (TDOA).  We are NOT in that realm here - speed of light is SO fast, that in a single tick, light has moved more than 6 meters (3e8/(1/48e6)) > 6 meters.  So I think we can totally forget about time-of-flight arguments.

So although we are looking at timings, none of them are to do with "it arrived here later because photons spent longer in flight".

**Aim -  we want 5 things:**
-------------------------
We want 5 pieces of information: bearing and elevation of the headset from the lighthouse (angles), distance between them (meters), and the orientation of the headset (where would the "wearer" be looking if they were looking straight ahead") (both are angles).  [ so 4 angles and a distance ].

Bearing and elevation of headset relative to lighthouse (infos 1+2)
-------------------------------------------------------------------
angular location (ingore orientation) of the headset relative to the lighthouse.  If the headset is "upstream" in the sweep, every sensor registers a "smaller" TSSP, and if the headset is "downstream" in the sweep, all the TSSPs are bigger.  If we average ALL the sensors TSSPs on a given sweep, we'll get a good idea of where the headset is (angle L/R in the horiz scan or angle up/down in the vert scan).  This measurement ignores the orientation of the headset, but is probably a really good approximation of the "angular position" of the headset relative to the lighthouse.  It doesn't give us the distance between them, though (time-of-flight of photons would be useful here, but they
are too fast).

Summary: By averaging the TSSPs in the H (V) scan, we get a reasonable longitude (latitude).  This is from the point of view of the light house (but easily reversed except be careful, we don't know the
orientation of the headset yet, so don't do that).

Orientation of headset, and it's distance from lighthouse (infos 3,4,5)
-----------------------------------------------------------------------

As a horizontal (or vertical) scan sweeps over the headset, it will trigger the sensors in a "wave" across the face of the device. (I assume the sensors to be on a single plane - perpendicular to the nose
of the wearer - like a face-mirror would be orientated).  This is an oversimplification but stick with it).  Imagine the headset at a random angle, and imagine a thin red light sweeping across it at
constant speed, triggering each sensor as it passes.  This can be thought of as a "wave".  The direction of travel of the wave, within the "plane" of the sensors can be measured from the TSSDs.  The speed of the wave can be calculated too, from the TSSDs.

The direction of this "wave of activations" gives us the 2 angles of orientation of the headset (the arrow points away from the lighthouse).

Finally, the _speed_ of the wave across the sensors will yield the distance of the headset from the lighthouse. (You have to take into account the angle of the headset, but we just calculated that above).

Note - all this needs only one lighthouse.  The estimation of R (distance between lighthouse(s) and headset) is much more stably calculated from the "triangulation" between 2 lighthouses.  I expect
everything improves lots with 2 lighhouses, but I haven't given it any thought.

**Comment by Zmathue** [comment]
So the tracking data is missing the length of the pulse, this is needed to recover a first pass estimate of the z coordinate of each sensor relative to the lighthouse, because the time of the pulse is roughly proportional to the inverse of z we can calculate it.

This z value is needed for first to resolve the ambiguity of building a 3d point out of a 2d projection (how do you know which way you are viewing the object from, if you view the object from the front it will appear the same as if it is viewed from the back) and from there it can be used to match it with the config data.

**Comment by cnlohr** [comment]
Hmm... I had no idea that was used at all.  You can tell which way you are oriented because of physical limitations, i.e. lighthouses above headset in Z... and because the curvature of the headset will give away any such failures. ALSO Don't forget I can get you the normals, so unexposed photodiodes will not react...

All that said, sounds like I need to expose pulse length!

**Comment by cnlohr** [comment]
I will take more data soon, but, something looks wrong with one of timings, I think I'm syncing to the wrong pulse.

**Comment by cnlohr** [comment]
Ok, here's a new one with lengths.   Also I corrected some more correlation errors.
[third_test_with_time_lengths.csv.zip](https://github.com/cnlohr/libsurvive/files/650897/third_test_with_time_lengths.csv.zip)

**Comment by cnlohr** [comment]
One other note... My math just isn't working out if I assume 400,000 ticks is the time of a circle.  

```
float angle = (hmd_point_angles[k] - 200000) / 200000 * 3.1415926535/2;  //XXX XXX WRONG??? OR SOMETHING??? WHY DIV2 MAKE GOOD?
```
I am just confused.  Anyone else got a chance to look at the data?

**Comment by mwturvey** [comment]
From what I've seen, the laser rotors sweep at 60 rotations per second.  Assuming 48mhz clock, that's 800,000 ticks for the time of a full circle.

**Comment by cnlohr** [comment]
Splendid, looks like I can go to sleep!  I still have a fair bit of work that needs to be done, like righting things so down is down, but this awful awful method might actually work!  (P.S. Located in 'tools' under 'planetest')

**Comment by Ardustorm** [comment]
Here is what I have so far. With what I had said above I was able to construct a surface that represents all possible points that an observer can be given two points and the angle (from observer) between them.
This results in a torus shape. I'm not too sure how useful this will be in practice since the points are so close together it is practically a sphere but who knows.
I was originally thinking that you could find the intersection point between a few of the torus' but since they would be so close together, I'm not sure that is feasible. It still might be helpful for other uses though.
[torus.scad.zip](https://github.com/cnlohr/libsurvive/files/652456/torus.zip)

```c
module torus(p1,p2,angle) {
   z =dist(p1,p2) /2; 
   r= z/sin(angle);
   R= r*cos(angle);
   e=(p2-p1)/dist(p1,p2); //end vector
   
   translate(p1)
      rotate( [0, 1*atan2(sqrt(e[0]*e[0] + e[1]*e[1]),e[2]), 1*atan2(e[1],e[0])])
      translate([0,0,z])
   for(th = [-180:10:179]) { for(phi = [-180:10:179]) {
	 x= (R +r*cos(th) ) * cos(phi);
	 y= (R +r*cos(th) ) * sin(phi);
	 z= r*sin(th);
	 color("grey",.8)translate([x,y,z]) rotate([0,0,phi])
	    cube(5.2,center=true);
      }}
}
```

**Comment by jpicht** [comment]
Can you post the full & correct JSON-Config file? I think for a neat solution we will need the model points and the model normals.

**Comment by cnlohr** [comment]
Here's mine, I've marked the sections that would need to be customized by someone's with XXX's.
[LHR-B4ABXXXX-Charles.zip](https://github.com/cnlohr/libsurvive/files/652896/LHR-B4ABXXXX-Charles.zip)

**Comment by cnlohr** [comment]
Here's the normals in CSV form for anyone who wants it...
[hmd_normals.csv.zip](https://github.com/cnlohr/libsurvive/files/652921/hmd_normals.csv.zip)

**Comment by Galastorm** [comment]
@cnlohr I think I follow what you're doing in the planetest: you're working on a single LH set of figures and checking x/y/z/ positions in a sphere around the cluster points to find the converge on the best sphere position (and thus angles) which fits the data?  But I can't figure out if you're taking the relative 6DoF of the LH emitters in to account?

For example each LH will have a different pitch/roll figure which would alter the XY angle ratios and therefore the XYZ positions worked out from those angles would not be using the same co-ordinate frame.  Imagine worst case of a LH on it's side...

(Please add a few more comments for those of us that need a little hand-holding! :-) )

**Comment by cnlohr** [comment]
I willl, I will! So, the way I was doing it was going pitch, roll, yaw, one at a time trying to find the best orientation.  This simply will not work at the macro scale... at least reliably.    It's also way slower, because I have to try a ?thousand? orientations before I figure out where it should be pointed.  I have to do that once for every position I try the lighthouse at.

It does happen to get the right answer most of the time for me, but this is bad.  I am going to try a more analytical solution, to purposefully point the lighthouse at the HMD from every possible position.  I will call this "planetest2"

**Comment by cnlohr** [comment]
Basically I was just trying every possible combination of position and pitch yaw roll I could think of.  This is bad.

**Comment by Galastorm** [comment]
Just to pitch in my two cents, this would be my starting point as to how I’d approach it…
Do the calibration dynamically rather than off-line, the user will expect to have to spend some time doing it and you can use more CPU intensive approaches within it that don’t have to appear within the standard mode.
Use the HMD or controller (in plain sight of both LH bases) to determine and then eliminate differences in the roll aspects of the LH bases first (standing in sight of both bases and PRY the headset whilst trying to keep the HMD in the same physical location).  Once you know the declarations are plumbed you should, I think, be able to start looking at the X and Y declarations as entities in the same plane.

Then move the controller up/down and forward/back all the time in view of both LH such that there is always at least one sensor that gets readings from both LH bases.  This should give you a whole load of points that you can do a simultaneous equation on to work out the relative pitch and yaw of each of the LH.

Finally move the controller in PRY and forward/back/up/down.  Take readings for pairs of sensors to look at the angular separations vs known maximum physical separations – you should be able to get a very good indication of the actual distance (and, with defined bearings to single sensors - the XYZ positions) by scaling against the defined HMD separation.  Unfortunately perspective will be the downfall of accuracy here because you’re never going to precisely match the bearings and the ratios.  

A much better solution for the final step would be the dreaded PnP to get both at once and make it a simultaneous equation…

**Comment by Galastorm** [comment]
Interestingly, the VIVE system also has the ability to refine the relative sensor positions from the default drawing positions on an HMD and controller when in use.  This removes the manufacturing tolerances from the devices and makes the overall system much more accurate in depth.

**Comment by cnlohr** [comment]
What you've described would be very cool, but, exceedingly difficult.  And, I would worry the absoluteness of the lighthouse system would be compromised.   I guess I don't fully understand what it would be solving, absolutely.  It would be nice to know the lighthouse locations better, and I think you could resolve better over time, but I think it would be a very, very long process, realizing the data consistently pulls one way or another.

**Comment by octavio2895** [comment]
I think I've found something that might help, sorry in advance if it doesn't help. I would try to program it but I'm a total noob.

OpenCV have algorithms that can approximate very fast the pose of a model given the 3D points from the model (the position of each sensor relative to any point in particular, say the center of the HMD) and the 2D points from the still image. This returns the translation vector (x,y,z) of the camara from the center of the HMD and the rotation vector (roll, pitch, yaw) from from the center of the HMD.

http://docs.opencv.org/2.4/modules/calib3d/doc/camera_calibration_and_3d_reconstruction.html

**Comment by cnlohr** [comment]
@octavio2895 : There is a lot of detail there when using cameras, etc... which thankfully, we don't need.  The lighthouses are all already pretty solid.   The other problem is that it's all so tightly coupled to OpenCV, itself which I would like to avoid inclusion of.

everyone else: I now have a tool able to determine location of the lighthouses pretty? accurately.  I verify the position by pretending to shoot rays out from the lighthouse where I expect the photodiodes on the HMD and see if it lines up.  Currently works for two lighthouses at the same time.  I really need to get issue #1 resolved.  I am sure it can be done better and faster than the way I am doing it.  See it in the tools/planetest2 folder.

**Comment by cnlohr** [comment]
Another test file if you guys want to try!  [testfour.csv.zip](https://github.com/cnlohr/libsurvive/files/656417/testfour.csv.zip)

**Comment by LeeCookGHA** [comment]
@cnlohr Would it be possible for you to create a test file with movement along the lines of the routine I outlined above?  I've just had delivery of a 300Mz Cortex M7 board  (double precision FPU and DSP extensions) and I really would like to see if I can't push through a PnP solution on to it.  I've some ideas about optimisations because the lighthouse angles are so accurate wrt to a standard camera.
Also the what are the 6 axis of the IMU?
No major hurry as I'm about to head to Bruges for the christmas markets for a few days!! :)
Thank you!!

**Comment by cnlohr** [comment]
So you want me to pick up the controller, move it around left right up down as I try to make sure the same sensor is visible to both lighthouses?

P.S. I can't right now since as soon as I power on the watchmen the data gets totally borked going to the headset.  Not sure why, though :(.

**Comment by LeeCookGHA** [comment]
@cnlohr Thank you :) Seriously though, there's no rush whatsoever. I'm going to be in no position to use the data for days, and possibly not until the holidays.
I can't believe that I am going to miss another live stream....

**Comment by LeeCookGHA** [comment]
Actually, it's the HMD that you have the position and normals for? Could you do it with the HMD please?
Thanks again!

**Comment by cnlohr** [comment]
Are you sure you typed what you meant @LeeCookGHA ?  If so, I don't understand your request.

**Comment by octavio2895** [comment]
@cnlohr He's asking for you to dump data when you are moving the HMD instead of the controller since we have more about the HMD than the controller.

**Comment by cnlohr** [comment]
oh! Gotcha! I think I will collect a variety of data.  Probably on the livestream tonight.  I'll see if I can get some now.  I do want to try capping data from the watchmen, too.  I will fix my one bug in the disambiguator.

**Comment by cnlohr** [comment]
@octavio2895  @LeeCookGHA  Here's the data of the HMD being still and the HMD moving.

[hmd_moving.csv.zip](https://github.com/cnlohr/libsurvive/files/659088/hmd_moving.csv.zip)
[hmd_still.csv.zip](https://github.com/cnlohr/libsurvive/files/659087/hmd_still.csv.zip)

**Comment by ErrJordan** [comment]
Nairol who did LighthouseRedox was in the chat and linked this but it was missed.

https://github.com/nairol/LighthouseRedox/tree/master/data/calibration%20data

This will get you controller calibrations.

**Comment by cnlohr** [comment]
This is the HMD static in the middle of the room. From the livestream 4.
[livestream_test_1_with_hmd_in_center.csv.zip](https://github.com/cnlohr/libsurvive/files/659215/livestream_test_1_with_hmd_in_center.csv.zip)

**Comment by MrRedBeard** [comment]
"computation of the Lighthouse positions"
https://trmm.net/Lighthouse

**Comment by MrRedBeard** [comment]
Reading lighthouse usb data
https://github.com/nairol/LighthouseRedox/blob/master/docs/USB%20Protocol.md

**Comment by cnlohr** [comment]
@MrRedBeard :: The github reference does not show anything about lighthouse receiving light data.

**Comment by MrRedBeard** [comment]
Ah I should have read in more detail.... multitasking

**Comment by Big-Kate** [comment]
from Sharp Labs  Correction of visible mura distortions in displays by use of flexible system for memory resources and mura characteristics  http://www.freepatentsonline.com/y2009/0096729.html
“Mura” defects are contrast-type defects, where one or more pixels is brighter or darker than surrounding pixels, when they should have uniform luminance... Mura defects may also be referred to as “Alluk” defects or generally non-uniformity distortions.
In general, most mura defects tend to be static in time once a display is constructed. However, some mura defects that are time dependent include pixel defects as well as various types of non-uniform aging, yellowing, and burn in"

**Comment by shaise** [comment]
don't know if it helps, but the following C code can reconstruct location/orientation of an object relative to a camera, if the object's points self 3d configuration is known, and the location of the points seen by the 2d camera given. This looks exactly as the case here.
http://users.ics.forth.gr/~lourakis/posest/

**Comment by cnlohr** [comment]
@shaise would you mind running it with some of the example tools?  You may be able to copy-and-paste the loading function at the bottom of camtest.c and it will give you the points on the HMD and vectors from the R or L lighthouse.

**Comment by LeeCookGHA** [comment]
Apologies if this is rambling it's late and I have just got back to to my hotel.
I don't think that you are using the HMD normals to best effect. If you look at the data shhet for the photo diode it has a fairly good relationship between the angle of incident and the power utpout, couple that with the

**Comment by LeeCookGHA** [comment]
actual time the laser will sweep over the diode surface and the times the sensors are lit wrt each other should be a reasonably weighted output as to the position of the lighthouse wrt the hmd normals.

**Comment by cnlohr** [comment]
I will give that a shot in camfind.c tonight.  I would imagine it should be related to the dot product.  Just wondering which photodiode is being used?  I was planning on using https://www.digikey.com/products/en?keywords=1080-1366-1-ND

**Comment by Zmathue** [comment]
bpw34  is the photodiode, and the TS3633 is used for amplification and envelope detection

**Comment by cnlohr** [comment]
Hmm, just saw the Upload article on it.  I think it's still worth me poking around on my own even with that, though that opamp is almost an all-in-one.  I wonder why the BPW specifically?  Not SMT, or is there an SMT version?  eeehhh... we need to take this offline, back to trying to use normals to refine estimates.

**Comment by geekmaster** [comment]
This suggests that the time between exposure of different sensors (at known distance apart) can determine distance, rather than the time of exposure of one sensor:

https://github.com/ShadowApex/liblighthouse-mocap

https://github.com/ShadowApex/liblighthouse-mocap/blob/master/examples/distance_3.gif

It seems that overlap of multiple normal frustrums could also help localize lighthouse location in a similar fashion.

**Comment by mwturvey** [comment]
For what it's worth, the TS3633's prototyping module (TS3633-CM1) comes with a PD70-01C/TR7 photodiode.  It's also worth noting that the TS3633 is the part that filters out signals that aren't modulated between 1-5MHZ.  While I haven't tried to intentionally inject interference, I've not seen any spurious signals from the TS3633-CM1, which has been helpful.   I wonder how much effect different photodiodes will have on measured pulse durations, particularly when hitting it at different angles.

**Comment by cnlohr** [comment]
@geekmaster I am including the length of the pulse in the data files, so hopefully that'll be able to help with the normals, too!

**Comment by LeeCookGHA** [comment]
Apologies for the drunken spam!
So what I was thinking was to look at the durations of the lit sensors, weight them based on the difference between the min and max (say applying 5% for min 95% for max) by applying a curve which is a of combination the responsiveness of the diode vs angle and the perspective of the diode surface wrt to the lh.
It needs to be a min:max ratio though as the absolute duration will be more a product of the distance than the angle.

**Comment by LeeCookGHA** [comment]
Two days and I'll be able try this myself.. Hopefully the summing of the weighted vectors will give a reasonable approximation of the direction of the lh.
Then, it should be possible look at the direction and angular ratio of the two sweeps over three sensors in a triangle. That should give a reasonable and fast approximation of the roll of the lh wrt to the hmd.
I'm hoping that it will be enough of a starting point for some of the pose estimation routines.

**Comment by shaise** [comment]
@cnlohr ,

I wish I had more time to work on it, its a cool project, but caught me just before going to 50 days vacation...
Anyway, I decided to do at least a quick and dirty test to determine HMD location relative to lighthouse.
I think the results are promising, since I randomly selected 4 sensors to do the math and the result (rotation and translation matrices) where fairly close.
The method to do that is using POSIT algorithm: http://www.cfar.umd.edu/~daniel/daniel_papersfordownload/Pose25Lines.pdf
I had no time to implement it myself, but it should not be that hard. Instead, I used a ready made c# LGPL library, Aforge.NET: http://www.aforgenet.com/articles/posit/
The attached sample c# program (should work on mono as well) takes HMD sensor configuration (no need for normals) and lighthouse data. (zip contains files I used)
After loading config and data file, a list of sensors are shown (for the L lighthouse). if you select any 4 of them you can do the calculation and get a rotation and translation (in mm) of the HMD relative to lighthouse.

Example result for sensors 10, 11, 12, 27:
Rotation Matrix
-0.023	0.392	0.920
0.994	0.108	-0.023
-0.108	0.914	-0.392
Translation vector
-541.042	-280.735	3620.589

Example result for sensors 8, 10, 11, 12:
Rotation Matrix
-0.038	0.428	0.903
0.998	0.068	0.002
-0.060	0.901	-0.430
Translation vector
-535.809	-283.032	3607.187

Just a few mm change and a 3.6 meters distance looks very good.

The estimator works on 4 points, but the original algorithm can work on more, so maybe it can generate better estimation. If you look into AForge sources, it looks rather easy to follow the algorithm.

Here is my source with .net binaries:
https://drive.google.com/open?id=0B3uO2iX6FayBRHlWUEtVVUxRUUk

**Comment by cnlohr** [comment]
@shaise : This looks like it's doing something similar to what I'm doing in camfind.c?  Or rather what I tried doing in camfind.c first... Then I found out I could be more intelligent about how I could be with fixing rotations... though this seems to be much better at converging to a position solution.  Also, there are a lot of other steps I hadn't considered that may aid its ability to resolve to an answer.  What data set was the one you ran from?  I think I am going to include some reference data sets in issue #8 - that way we can start using common data.  First I need to fix the disambiguator for the watchman controllers.

I will say your data looks like it's predicting the lighthouse being too far away.

@LeeCookGHA : I wonder if we'll be able to make something like camfind more generalized.

**Comment by shaise** [comment]
I was using correct_hmd_points.csv and livestream_test_1_with_hmd_in_center.csv L lighthouse. These are included in above zip file. If you have a windows system (or Linux mono) to run the binaries, you can try other data.
I don't know if it is exactly the same as you have done before, but if so, did you project the angle values onto a 2D plane before calculating the pose?
I wish I had the time to convert all code to C and ditch any external library. I can still do it 50 days from now if needed. The nice thing about the POSIT article that they claim its good for realtime 60fps tracking as demonstrated on a 1992 MAC...

**Comment by cnlohr** [comment]
I might be able to give it a shot myself.  Conveniently, this process of finding the lighthouse does not need to be realtime.  Once the lighthouse positions are found we can proceed to the next step. I do not mind re-coding things myself too much, either.

"Going on a 50 days vacation" :-p vacations for me are working on projects, so if I was saying something like that, I'd have unlimited time to do this!

**Comment by ElectronAsh** [comment]
Hi,

I'm just having a look at the raw data atm, and trying to determine what can be gleaned from the "step" values alone.

(I've been calling the timing data the "step" values, but I'm not sure what your terminology is atm?)

So, for example, I'm looking at the first 15 data points for the X sweep in "newtrackdata.csv"...

[http://imgur.com/a/ozhrA]

I'm just trying to get my brain around the basics atm, so bear with me. :p

I'm guessing that the current algo for the camfinder is basically doing something similar, but I'm looking to calculate what the difference value would be for when two of the sensors are perpendicular (and equidistant) to the sweep?

So, for the first two values of 190,444 and 190,649, the "step" difference is 205, but what would that value be if both sensors where exactly perpendicular to the laser sweep?

I imagine this is basic trig stuff, but I was never great at that at school. lol

I'm thinking that you might only need to do an interative calc for only each "strip" of sensors at a time (one calc for the X sweep, and one for Y.

Also, I'm assuming that the master clock is always running at 48 MHz for the sweep "step" counters, and that's what gives us the 0.0009 degrees per step?

Plus, the sweep period is essentially only 8.333 ms for ONE of the X / Y sweeps, since it essentially only covers 180 degrees of usable area after the sync pulse.

That also means that there are a maximum of 400,000 counter "steps" for that 180-degree sweep?
(800,000, if we're counting the full 360-deg rotation of each rotor?)

Each rotor spins at 3600 RPM, correct? So 60 "sweeps per second" for the X rotor, and 60 SPS for the Y rotor (but alternating, of course)?

If I could just calc the step values for each pair of sensors being "square on" to the sweep, and then at almost a 90-degree grazing angle, that would be a good start.

Obviously the timing data for the sensors is already sorted in order of the edge detection, but I've also sorted the HMD sensor point data on that spreadsheet, and made it so the "sensor distance" values are always positive.

I've installed OpenSCAD as well, but I can't seem to get it to display properly if I modify the code and then re-render? It only seems to work after I first run the proggy, then open a file?

Anyway, the idea is that once the step value differences (between any two sensors) have been translated into the rough angle of the incoming laser sweep plane, every consecutive pair of sensor values would help to improve the accuracy of the resulting angle.

Then the same would of course be done for the Y and Z "distances" between the sensors, and then everything combined into one vector.

I'm not sure how the sensor normals would work with all of this yet though.
I imagine they are extremely important for both accuracy, and determining if the incoming beam is further forward or further behind each sensor pair, but would just the step difference alone give a fairly good estimate of the LH position?

I think I need to order a few nice IR sensors, and build a rudimentary LH.
I don't have a Vive yet, but I'm definitely interested in getting one eventually.

**Comment by shaise** [comment]
Actually, maybe i'm wrong, but I think with this algorithm you do not need to find the lighthouses. You can apply the same algorithm to get translation / rotation of the controllers. All done on the fly per dataset. The controllers values will also be relative to the HMD if you diff the matrices. So all that is needed now is to wear the HMD in mid room, face a default direction in the room, and set this translation/ rotation as a calibrated base. 
Regarding vacation, I can safely say I have all year round for projects - SW HW and mechanics (as I work only few hours a month), but once in a while need to have some network challenged nature around...

**Comment by ElectronAsh** [comment]
lol

I just skipped into the last live stream vid again, and at 2:12:00, Mr Lohr explains how he's already converting the LH data into radians for each sensor.

Is it not enough to just use those values to narrow down the angle which points TO the LH, then do the same against the Y and Z sensor point values (and the Y sweep), then combine the X,Y,Z angles into a single orthogonal vector?

Once you have a vector pointing towards the LH, and then calc the vector to point towards the other LH, would that not give the point where the two vectors cross, and thus the HMD position?

(I know, that's basically what you're doing already, but again, I'm just wondering if I'm way off-base here?)

**Comment by shaise** [comment]
I think this was also hinted by @vk2zay's mode 'A' - one lighthouse. There is no actual need for 2 lighthouses as one is enough for positioning, the reason for using 2, is for the cases where you turn around and your body obscures the controllers from lighthouse A.

**Comment by ElectronAsh** [comment]
Yep, and I read an article last night where they said that the positioning jitter (for the HMD at least) is only 0.3mm when two LH are used, and still only 2.1mm with one LH. Pretty damned impressive. :o

I really should buy some sensors to mess with now, as I could at least implement the basics of this on an FPGA, and get very accurate timing data back (using the same 48 MHz clock).

It will be cool just to see how the timing difference changes between a pair of sensors when their angle is changed relative to the sweep.

I'll have to buy some laser line gen lenses etc. I have plenty of spare laser diodes, at least. :p

**Comment by ElectronAsh** [comment]
I'll have to try compiling this code.

I'm amazed I can actually read and understand it.

I didn't even know you could do something so simple to grab the values out of a CSV.

float fa, fb, fc;
int r = fscanf( f, "%g, %g, %g\n", &fa, &fb, &fc );

...Where %g is the format of the number in the CSV (floating point), \n is the newline char, and it just spits out the values to the address pointers of the floats. ;)

I'm guessing that fscanf returns a number for how many requested values it found?

**Comment by cnlohr** [comment]
@ElectronAsh Yes, indeed we have the angles from lighthouse center for the points -- though I am thinking we should probably sanitize them with something to throw out outliers.

@shaise Are you referring specifically to using the normals of the incident light on all of the sensors on the HMD?  Or are you referring to just getting the angles from the lighthouses' POV?  

@ElectronAsh I don't believe it's possible to get relative angle from one lighthouse to the other.  If there is a way of getting precise timing of when one lighthouse sweeps the other that would be most dapper.  I just haven't seen any references to that.

@shaise I have heard vague murmurings about this being possible without knowing lighthouse location, however, I am most concerned with simplicity of tracking solution.  Calibration can take a while, that's okay.  Tracking cannot.  I believe if we know the locations of the lighthouses, the path to getting position in a global frame can become much, much easier.  I could be wrong in this belief, though.

@ElectronAsh I am seeing INSANELY good tracking of object position based on the lighthouses, but it seems to be much more difficult to determine lighthouse position.  The error I'm getting on some of my "best lighthouse positions" are in the order of 0.6mm per photodiode!  Unless we "know" lighthouse positions, this is not applicable, though.  Your laser line thing sounds like a hackaday publication in the making!

P.S. About the C thing.  I don't use C because I'm as masochist.  C does as much as is sensible in the language to make things as simple and easy as possible.

**Comment by ElectronAsh** [comment]
I like C.

It makes the most sense to me as a coding language (although I'm still learning).

The C++ extensions just confuse the hell out of me.
Like when they changed the whole MAME / MESS source from pure C to C++. So many macro everywhere, and code spread across more files, which makes it far more difficult for me to follow the code in a "linear" manner.

I find most C sources as far neater and easier to read, while still achieving 99% the same thing.

**Comment by cnlohr** [comment]
Back on topic: I think I'm going to try to switch camfind.c into polar coordinates.   Then, maybe find the centroid of the precision distribution.  Also, still need to get the disambiguator working better with the watchman controllers.

**Comment by shaise** [comment]
@cnlohr , I was referring to just using lighthouse POV angles and HMD sensor configuration. No normals. (not required by POSIT). I feel it is fast enough for realtime tracking (I was able to perform the entire computation 100000 times per second on my 3 year old i5 laptop, using the slow c#.)

Anyways, hope to catch at least some of the next live sessions on my trip through NZ. You guys rock!

**Comment by LeeCookGHA** [comment]
Feeling very lucky to be back from the Christmas markets in Belgium after the horrendous attacks in Germany!
I've just taken a quick look at the normals data vs light-up for a side-on view data file and it looks quite promising for being able to quickly provide an accurate initial pose to the LH. With the HMD on the floor (approx -2m down and 3.5m away?) that would put sensor 1 presenting an aspect ratio of around 1:0.84 x to y - we see timing ratios of 1:0.87.
It also looks like (through sensor 12) that there is a minimum amount of power necessary at the sensor in order to detect that it is being lit and that the laser width is likely much thinner than the sensor width (2.75x2.75mm if it's the Vishay part)).  With a very shallow side perspective it only lights in the X axis (i.e. virtical laser axis) implying that the coverage of the laser in the Y axis is not enough?
Will hopefully be able to look at more of the data to try and get a better feel for the problem variables this evening after work.

**Comment by LeeCookGHA** [comment]
Ok, so looking at it some more it's nowhere as promising for an accurate vector combination as I thought (I've deleted my previous post as confusing).
I pulled all the info from (livestream_test_2_x_axis.csv) in to a pivot and looked at the averages rather than a dip in to the centre, and the lit timings must be affected a lot more by the width of the laser line than I thought.  Assuming the photodiode sensor area is 2.75mmx2.75mm that means the laser centre is pointing at the diode surface for approximately 100 to 175 ticks at distances of 2-3m (60hz, 21600deg/s, 48M tick)  But there are sensors with averages of up to 557 ticks (RX30).
This implies that either the laser dispersion (6+mm width?) is lighting the well before the centre of the laser line reaches the sensor, or that the processing circuitry is somehow extending the duration, or a combination of the two.

More investigation is necessary I think...

**Comment by cnlohr** [comment]
@shaise : I think a lot of good data can be pulled from the normals and length of pulses.  Makes me nervous to try to use POSIT, especially since it doesn't look too different from what I tried initially, just that it would converge much, much faster.   Though that would mean I could supplant my original algorithm if I could figure out what was going on.

@LeeCookGHA : There is a fair bit of analog circuitry from what I understand that helps infer intensity, not just dispersion.  From what vk2zay mentioned in one of the livestreams when I suggested piping the thresholded data straight into an ESP32, he indicated that wouldn't be a good idea - so I'm thinking there is more to it than dispersion width.   This also kind of makes sense if you think about it as charging an integrator if it's very bright that will take several light-cycles to decay and finally go under the threshold.

**Comment by ElectronAsh** [comment]
I have a strong feeling we may get a few hints from this vid. So weird that it should show up right now...

https://youtu.be/75ZytcYANTA

**Comment by vk2zay** [comment]
It took Hackaday a little time to edit the videos, the talks have been showing up one by one every few days.  There is nothing much in that video about the inverse problem you are currently trying to solve, it is more about the hardware challenges.

I can confirm some of the assumptions in this thread though.  The width and divergence of the beam, the sensor size and its pose with respect to the base station is responsible for the varying pulse widths you see for rotor beam hits.  You will need to use the centre of the beam hit pulse, not its leading or trailing edges.  The exact width varies, the middle of it estimates the true centroid.

Your volumetric plots confuse me.  As you have noted yourself it appears angle is less well determined than range.  This is not true in general for triangulating systems; something is broken.  You might like to attack the problem in 2D first then extend your solver to 3D.  In 2D an object pose has only 3 degrees of freedom and the maths is simpler to reason about, but has basically the same form as it will be in 3D.  Starting with one beam, one sensor, then adding sensors until the solution is unique is instructive.  Directly trying to invert the projection problem may not be the best approach.  PnP and POSIT are built around camera models, they are instructive to study, but probably not the way you should solve the Lighthouse inverse problem.  Lighthouse is definitely not a camera, especially once you utilise the calibration information for the base station and absolutely not for moving objects.  That shouldn't stop you from trying simplified models on static data sets first, but for good dynamic performance with real-world base stations you will need something more complex.

**Comment by cnlohr** [comment]
@vk2zay: So, getting back into this now... After looking at some hypothetical situations, it seems that the scatter of being fairly precise as far as depth goes but not precise as far as lighthouse position goes is to be expected.  I suppose I should make this more concrete, and run my voxel representation on more "perfect" scenarios.  For that aspect of finding the lighthouse, I think switching to polar coordinates would be best.  For me, I think it would be more difficult to do 2D then extrapolate to 3D.  I work much more intuitively in 3D.   Additionally, I am specifically only interested in two-lighthouse environments, so it makes sense for me to stick to that especially considering approach 3 listed below.

I'm definitely discarding POSIT and PnP at this time.

I do want to do some more things to try to refine the solution.  I think coming at it from multiple angles and using these several pieces of information will help resolve to the best solution possible.  I intend to:

(1) Try shifting answers by the length-of-pulse up and down some to verify I am using center-of-pulse.

(2) Measuring pulse time as a function of distance and dot-product of the lighthouse-normal.  Then using this to aid the lighthouse position resolver.

(3) (Dependent on #4 and #11 ) Use watchmen controllers in role of evaluating positions of lighthouses, but not just to determine which way is "down" and the "floor" but also to resolve it from the approach of:   If lighthouse B is here and lighthouse C is here, they both make sense for the HMD, but their expectation of where the controller is makes no sense.  My only concern with that is it would require at least one photodiode on the watchmen to be visible to both lighthouses.

(4) Just toying with it: I could skip from this (Phase 2) of my grand libsurvive scheme straight to Phase 4 (Making an ESP8266 sensor) and ignore Phase 3 (Finding locations and poses).   I would just make a jig to calibrate the lighthouse locations with by using a coordinate frame.  Of course, this jig was for a camera localization problem I had to do a few years back, would be using photodiodes.  But it would by golly make it easy to define a coordinate frame for the room and find the lighthouses.

![dsc01494 - science time with glowy ping pong balls for localization](https://cloud.githubusercontent.com/assets/2748168/21461397/3da4ae9e-c91f-11e6-9405-a67a95af4ee2.JPG)

**Comment by nairol** [comment]
My idea for finding the location and orientation of a base station would be using the Inscribed Angle Theorem in combination with the Law of Sines and other basic geometric formulas. I've mentioned this in chat in the last live stream but without explanations it's pretty hard to understand. (Remember the PNG file with the triangles and circles?)

I've made a video explaining the concept in 2D but I don't know how to solve the problem in 3 dimensions yet. Should be possible somehow™.
Any math/geometry people in here that like a challenge? :)

[Video](https://youtu.be/EgDfKvzdMP0)
[Geogebra file](https://dl.dropboxusercontent.com/u/24771452/HTC%20Vive%20Stuff/Inscribed%20Angle%20Theorem.ggb)

Also sorry for the poor voice-over quality. I recorded this at 4 AM and am not that fluent in the (spoken) English language to begin with. Youtube's replay speed setting can help if you don't want to sit through it all. :)

So basically this would be a non-iterative approach to the problem but I'm not sure if it would work that well with Lighthouse's phase-shifted J/K sweeps when the tracked object is in motion. Might work in combination with an iterative optimization step afterwards.

**Comment by Sawaiz** [comment]
I attempted to process some of the data and attempt to graph it in 3D.

https://sawaiz.github.io/viveDecode/

This data is from the hmd_moving (hmd_still has some slight movements so this is a bit more interesting to watch). The data is split up into events (400k ticks for x and 400k ticks for y create a event) and then graphed in time. The x is converted to phi, and y to theta, and the radius is currently a constant.

Points to note
- The points at 0,0,0 and those stuck on the axis are the ones missing some data points, those should be removed from analysis.
- There is a timeout on the points that don't get hit, so those need to be filtered out. This can be seen by graphing time from sync vs time. Its likely some number of ticks from the pulse.
- There is some sort of mirror image on my graph, everything is from LH1, so I don't know where that could be coming from, perhaps my math is wrong somewhere.

I'll likely be trying to build the point constellation next to compare against the points, from three matching points in the constellation, you should be able to find out what he distance is.

**Comment by cnlohr** [comment]
I just saw this now.  I wanted to comment in this thread, I think I have the data from the HMD and watchmen going.  File is: find_lighthouses_with_watchmen.csv.zip in #8 

When I click that link above, nothing shows up.   Should I be seeing something?

**Comment by octavio2895** [comment]
@nairol I think we can get a faster solution using spherical coordinates and vector algebra. In 3D, for every sensor we know 2 angles **tetha** and **phi**. In order to reconstruct the sensors in 3D we are missing lenght or **r**. We also know the distance between each sensor lets name it **k**. First we grab the angles from 2 sensors, then we subtract vectors and let them be equal to **k**. We do this with 1 more sensor and we are left with solvable systems of equations (3x3). The problem is that we would end up with more than 1 solution, so we add another point to the equations. We now know the position of the object relative to the LH.

**Comment by Zmathue** [comment]
Every approach I have taken to this problem requires calculating the rotation matrix for lighthouse which has to use cosine terms to relate the 9 variables to the 3 Euler angles but this causes the solution to be nonlinear which means you then need to use an itterive approach to get a good solve, and yeah then you will have more data then equations so you need to use least squares which only further complicates things. But I have a couple more ideas to try

**Comment by cnlohr** [comment]
Check out the data for distance/angle to pulse-length in #8 - maybe we should be looking at a hybrid approach that uses multiple mechanisms and then we select the best based on several approaches.

**Comment by Sawaiz** [comment]
Something like this, I just tested chrome, edge, and mobile chrome, so It should work. Its 3d so you might need to zoom out or rotate a bit. 

![vive decode](https://cloud.githubusercontent.com/assets/6422947/21490233/54cd90ec-cbc0-11e6-908c-221cd0ab50ac.png)

I would like for the solution to be non iterative though, it just feels so much more elegant.

![vivetimesync](https://cloud.githubusercontent.com/assets/6422947/21490386/0a45c196-cbc2-11e6-8512-95129290c50c.png)

While I'm here, I'll also drop this in here, its the time vs time from sync. The points at the bottom in a straight line are points that are timing out I believe. Don't know if its much help, but It made grasping what was happening a lot easier for me.

**Comment by cnlohr** [comment]
@Sawaiz -- lol you implied edge was a browser. I used firefox (doesn't work).  It works in Chrome.   

I'm still not 100% sure what I'm looking at. If it is the HMD position from the PoV of each lighthouse?  If so, what tactic are you using?  Is it something more naive?

**Comment by mwturvey** [comment]
@nairol Thanks for the video description.  It's been long enough since I've had trig that I didn't recall the inscribed angle theorem, which is massively helpful. One thing I ran into is that you don't just have one circle that the lighthouse can be on-- instead there are two, depending on which direction you choose the line for the center of the circle.  Other may have already figured this out, but it struck me that with two points in 3D, the surface that the lighthouse could be on is a torus.  And that's why when you take a 2D slice, you get 2 circles representing where the lighthouse could like.  I was trying to figure out a way to calculate the intersection of tori, but maybe that's a bad approach.  Instead, if we can use three points to define a plane, then work on that plane to discover the two points where the lighthouse could lie, is there a way we could then rotate that plane to define a circle in a 3D space?  Not sure what axis it would have to rotate around, or if that's even valid.  But, if we can get multiple circles in 3D, that seems like a much easier intersection problem to solve than the intersection of multiple tori.

**Comment by cnlohr** [comment]
@nairol - my brother was thinking a similar approach.  I was toying with trying it, and just using an iteritve approach in 3D to try to find the optimal intersection, since in 3D, I think it is a donut shaped object that the inscribed angle theorem produces. (I think @mwturvey is right about it being a torus, since it's basically a circle wrapped around the axis made by the two points you're looking at.) 

@mwturvey I don't understand what you are talking about when you are talking about another possible solution wrt not intersecting multiple tori.  (I could have sworn it would be greek, so torodes, but as it turns out, nope, latin!)

**Comment by mwturvey** [comment]
@cnlohr I was trying to figure out a way to avoid having to find the intersection of multiple tori, but I think my comment above amounts to sleep deprived ramblings.

**Comment by mwturvey** [comment]
I'm approaching the problem with simulated data first.  Given 3 points and the relative angles from a simulated lighthouse, I can generate 3 tori.  In the image below, you can see that the lighthouse (red cross) is at the "intersection" of the tori.  I still need to work the algorithm to find the intersection, but at least I can see what I'm doing now.  

It's also interesting to see that there are two points where the tori converge-- both at the lighthouse location and in the exact opposite direction of the lighthouse.  I guess the normals will be helpful in filtering that out.

![tori](https://cloud.githubusercontent.com/assets/18567877/21550886/926098ea-cdba-11e6-9a07-0073e7dff6cb.png)

**Comment by geekmaster** [comment]
Wow. That intersection looks a lot like a portion of the "convergence" animation that Alan Yates posted here:

https://twitter.com/vk2zay/status/812220006129745920

**Comment by mwturvey** [comment]
Wow, I had no idea what to make of that animation when I first saw it, but, yeah, I think you're right!

I played with this a little more: I added a 4th point which now generates 6 tori.  I assumed the normals to all be in +Z, so I could be lazy and just chop off all -Z points as possible LH solutions.  I then took one of the tori (the light blue one), and for each point I found the nearest point on each other torus.  I used that to compute the least squares distance from each blue point to all of the other tori.  I then put a light blue star on the point that matched the best. (Look in the lower right.)  The red star is the location of the actual lighthouse.  It's very much worth noting here that I haven't introduced any noise into the signal, so not sure of the noise immunity with this type of algorithm.  I did try it with a much lower angular quantization, and it appeared to still work pretty well.  I can post pics if anyone wants them.
![findingthelighthouse2](https://cloud.githubusercontent.com/assets/18567877/21554534/9833ff74-cdcd-11e6-9c0b-4dbb47d5e09d.PNG)

**Comment by cnlohr** [comment]
Sounds like it's getting dangerously close to a time when you should be trying the real data.

**Comment by vk2zay** [comment]
@geekmaster in 2D one way to solve the problem is basically circle intersections, pick one of the three sensors, you know the distance from it two the other two.  Each distance is a circle where the other sensor could be if you pin down the first sensor.  Similarly for the 3rd sensor there is a circle (likely of different radius) from the first where it can be.  Now you can parameterise the distance of the 1st sensor along the radial from the base, this parameter can vary from zero (touching the base axis), out to the distance where the circle locus to the 2nd sensor intersects the radial to the 2nd sensor only at one point.  (i.e. for only 2 sensors there is an infinite number of solutions where the line between the 1st and 2nd sensors lays in the plane segment between the two radials).  Adding the 3rd non-colinear sensor with its own circle from the 1st (or the 2nd if you like) and the intersection with the 3rd sensor bearing radial gives you a unique solution for position and orientation.

In 3D it can be done the same way, but with beam planes and sensor distance spheres instead of lines and circles.

You can think of it as adding constraints for each sensor you have a measurement for.  For zero sensors the object can be anywhere at any orientation.  With 1 sensor all you know is that sensor must lay on the beam plane (let's use the camera approximation and say radial line from two beam plane intersections instead of just planes, but you get the idea...), it can be at any range, any orientation (well you can narrow it down to a cone of orientations with the sensor normal and its response angle, but you need not).  Adding a 2nd sensor means the line between the sensors must lay in the plane spanned by the two sensor radials but the object can also rotate around the sensor-to-sensor line by any amount (and rotate about the cross product of both radial directions as long as the sensors themselves are still on the radials).  Adding a 3rd non-colinear sensor lets you solve for a finite number of orientations per selected 1st range, but the overall solution is not unique, there are multiple solutions where different sensor ranges can produce the same bearings with different orientations.  Adding a 4th non-coplanar sensor will disambiguate the pose solution.  Overdetermined solutions using 4 or more sensors are best done with iterative methods that minimise the residual projection angle errors of the solution pose - perhaps using subset consensus methods to reject outliers if you have a large amount of redundant data.

The shape of the convergence region of the 2D 3-sensor problem is probably not related strongly to the 3D version.  In 3D you have 2 additional rotation degrees of freedom and only one more spatial degree of freedom; the expressions for the Jacobian elements are much more complicated because they have 3-space rotations,  Some 2D slices of the 6D map of the 3D problem might be similar to the slices of the 3D map of the 2D problem...  Guess I should try plotting some.

**Comment by octavio2895** [comment]
@vk2zay That's very similar to what I was trying to do, I figure that if you know the coordinates in the camera view and the coordinates in the world view then you could determine the orientation and translation of the HMD. Using spherical coordinates and the distances between sensors, its very easy to compute the position of each sensor in camera view. Then using a this method http://nghiaho.com/?page_id=671 we could calculate the rotation and translation of the HMD.

**Comment by vk2zay** [comment]
@octavio2895 yep, something like that should work.  The solution (pose of the tracked object) is the rotation + translation that minimises the error between the measured bearings and the bearings that would result at that pose.

Given model points Pn posed at a translation of x and a rotation of r there is a projection function F(Pi, x, r) -> (azimuth, elevation).  This allows you to formulate an optimisation problem; fit F(Pi, x, r) - Bi = 0 using x, r over i where Bi is the measured azimuth and elevation of Pi.  The function F is basically two parts, the posing bit which gives Pi' = R(r)Pi + x and the projection bit which turns Pi' into Bi and is of the form P(x,y,z) -> B(azimuth, elevation).  Putting the base station at the origin makes the projection easy.  The only thing which makes any of this nasty is that differentiating F gets ugly fast.

**Comment by mwturvey** [comment]
Instead of raw data, is there any way I could get some data for the angle/ tick count of the tracked sensors, as well as an approximate location of the lighthouse location relative to a static HMD?  (Basically the physical location info and output of the disambiguator, I believe.)

Eventually, it would be nice to have a few such data sets for testing purposes, including both having the HMD relatively centered in the lighthouse's "view," off to an angle, and also having the HMD facing toward the lighthouse and off toward a side.

I did plug in the data provided by @Ardustorm above, and got a lighthouse position of:
(0.380, 0.491, 1.875) for that data set, but I don't know where the lighthouse really was to compare against.

**Comment by LeeCookGHA** [comment]
So this was my idea for a slightly different approach toward getting lighthouse position.  I still haven’t worked through the maths to see if it is viable, but I think it should be.

Take three sensors which have light data.  Create a plane in the HMD reference frame from those three sensors.  Now create a circle on that plane which intersects those three points choosing one of those points to be a zero degree point (this point should ideally be common with other plane/circles creations?) and work out the angle of the other two points on the circle wrt to the zero.

For those three sensors take the X/Y (az/el) swept angle positions and create an ellipse.  Generally you need 5 points to create an ellipse but, by having the known angles it should be possible with three. 

Do this for as many sets of three sensors as is possible.

The idea would be to then fit those ellipses to their respective plane/ circle using cone/plane intersection algebra to give the position of the point of the cone.  The point of the cone for each set should be the same, and also be the origin of the lighthouse.

Thoughts?

**Comment by LeeCookGHA** [comment]
Thinking about it some more, because lighthouse gives absolute angles, I think it should be possible to lock the position and orientation of the HMD wrt to the lighthouse with only two sets of three – and potentially without the cone/plane fitting…

As before, create two circles with two sets of three points getting: the centre of the circles, the points with angles and, the normal of the plane the circle sits on.  Calculate the distance between the centres of the circles.

Create the ellipses as before locating the centre of the ellipses and work out the normals of the associated circles from the ellipses (this will be the circle normal in the reference frame of the lighthouse).

Lock the HMD plane normals to those created by the ellipses – this should lock the HMD attitude in the lighthouse reference frame.

In theory it should now be possible to align the centres of the circles with the bearings given by the centre of the ellipses to give a distance from the lighthouse.

Talking rubbish?

**Comment by mwturvey** [comment]
I don't have a Vive, so I can't gather sample data through libsurvive.  But I do have a pair of lighthouses and a custom made target with 6 sensors.  Using it, I captured 15 [sample data sets](https://docs.google.com/spreadsheets/d/1GNS33sofkdDuClAb6bnLx5ksAbjQ9mrovAtQ-Lhgskw/edit?usp=sharing).  I've run about half of them through [my code](https://github.com/mwturvey/lighthouse_sim) so far, and so far almost all look like they're converging nicely.  (There's one data set that is an outlier, and it looks like one of the sensor reads was skewed.) 

The biggest drawbacks right now to the algorithm I'm using is that it eats up a lot of RAM and compute resources to generate point clouds for every torus and then iterating through the points and computing distances.  However, there's still a huge amount of potential to optimize it down.  Also, I'm not yet attempting any calculation of the orientation of the lighthouse.  

I'd like to start cleaning this up and integrating it.  Where is the best place in the architecture of libsurvive to put lighthouse location detection?

**Comment by octavio2895** [comment]
I've been testing my code with data you've been sharing. It's not exact but that's because I'm not averaging nor applying any type of filtering to the points, I'm just using the first 4 points on every sweep and threat them as accurate. Surely there is a lot too improve, like developing an algorithm that grabs 4 random points of the HMD, process it and averages with any other 4 points. Or an algorithm that's able to distinguish points that are off directly (like any point that's too separated from all others can be excluded without any analysis), indirectly (like using the IMU to estimate the pose between sweeps) or analytically (like analyzing all poses from random points and determining which point is causing error, or using the normals with the average pose and determine which point is nonsense). By using my code and the data As-Is yields the following [results.](http://pastebin.com/YB40YLLS)

By comparing my results to the measurements you gave, I get an accuracy of about 5-10 cm which I expected. My guess is that by manipulating the data with the methods I described earlier you can get closer to reality. The Euler angles are in degrees and the translation vectors are in meters, I'm not really sure which distance is XYZ and which angle is roll pitch and yaw because is taking the point 0, 0, 0 of the HMD as the origin (so it depends heavily on the angle) but my best guess is that it's YXZ. The RMS error you see is how close is the the pose reconstruction given the distance of each point, it's not a good way to measure the accuracy of each pose because its not considering the accuracy of the distance. I don't expect it will get any better than this.

I'm cleaning up my code to upload it asap. It's in python btw (sorry?).

**Comment by LeeCookGHA** [comment]
I've been implementing the idea that I outlined above and, from what I can tell, it works well and is quite fast:

https://hackaday.io/project/14977-room-based-vr-positioning/log/52081-about-turn-the-mathematics-done

**Comment by octavio2895** [comment]
[Here's the code](https://github.com/octavio2895/lighthouse_pose) I used, it solves the pose at about 130-150 Hz, there's a ton of room to improve.

**Comment by cnlohr** [comment]
@octavio2895 I have tried your code.  I'm not sure where your source data came from, but I tried using the data from issue #8 and, I'm getting much larger errors: 
```
Translation Vector
[[ 0.30292643]
 [ 9.8524632 ]
 [ 7.24489426]]
Rotation Matrix
[[-0.89377251  0.44851961 -0.00092523]
 [ 0.25054407  0.49755022 -0.8304646 ]
 [-0.37201931 -0.74247824 -0.55707064]]
Euler Angles
-153.386804368 123.853453758 179.936165985
Distance [[ 12.23320434]]
RMSE: 0.0941585319627
```
It is definitely not 9,7 units of distance away!  Any way you can hop on discord today and we can work through it?

@LeeCookGHA - Will start reviewing your work now.

**Comment by cnlohr** [comment]
@mwturvey -- will be reviewing yours too.  I'll be in discord for the next bit, as well as later tonight.

**Comment by octavio2895** [comment]
@cnlohr were you using the data with watchmen? My code does not distinguish between hmd and watchmen at the moment, its an easy fix though.

**Comment by cnlohr** [comment]
I included WM0/1 in the data, you'll need to throw it out.  I think to get really good solutions you'll have to use more than the HMD, but that's just a guess.

**Comment by cnlohr** [comment]
I think we can finally put the nail in this issue's coffin.
